### PR TITLE
Coach can create session

### PR DESCRIPTION
**Pull request by johannaanden**
# PT Link

[Coach can create session](https://www.pivotaltracker.com/story/show/163531630)

# User Story

```
As a coach
In order to utilize my spare time
I would like to be able to make sessions on Tribal
```

# Screenshots

feb-05-2019 13-03-20

# Description of PR

- Adds scheduling for creation of session

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+class SessionsController < ApplicationController
+
+    # before_action :get_session, only: [:show, :index, :create]
+
+    def index
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove the index action

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        @sessions = Session.all
+    end
+
+    def new
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove the new action

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        end
+    end
+
+    def show
+        @session = Session.find(params[:id])
+    end
+
+    private
+
+    def get_session
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove this method

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        @session = Session.new(session_params)
+        @session.save
+        if @session.persisted?
+            redirect_to root_path, notice: "Session successfully created"
+        else
+            render json: {errors: 'Every field needs to be filled in!'}
+        end
+    end
+
+    def show
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove the show action

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    def new
+        @session = Session.new
+        @price_tables = PriceTable.all
+    end
+
+    def create
+        @session = Session.new(session_params)
+        @session.save
+        if @session.persisted?
+            redirect_to root_path, notice: "Session successfully created"
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove the notice

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+class SessionsController < ApplicationController
+
+    # before_action :get_session, only: [:show, :index, :create]
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Remove this commented out code

File: app/javascript/css/tailwind.scss [hunk 7]
```diff
@@ -61,6 +61,9 @@
  * @import "utilities/skew-transforms";
  */
 
+.whitetext {
```
Comment (GergKllai1) [inline app/javascript/css/tailwind.scss]: Why are we having this class?
You can use the tailwind `text-white`

File: client/bundles/Main/components/sessions/AllSessions.jsx [hunk 8]
```diff
@@ -7,7 +7,7 @@ const AllSessions = props => {
 	let sessions = sessionlist.map(session => {
 
 		return (
-			<div className="wrapper">
+			<div className="wrapper" key={session.id} >
```
Comment (GergKllai1) [inline client/bundles/Main/components/sessions/AllSessions.jsx]: Why you added this key?

File: config/routes.rb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   devise_for :users, controllers: { omniauth_callbacks: :omniauth_callbacks }
   root controller: :main, action: :index
+  resources :sessions, only: [:index, :new, :create, :show]
```
Comment (GergKllai1) [inline config/routes.rb]: should be only create

File: config/routes.rb [hunk 10]
```diff
@@ -1,5 +1,7 @@
...
   get '*path', to: 'main#index', via: :all
+  post 'sessions/new', to: 'sessions#new'
```
Comment (GergKllai1) [inline config/routes.rb]: remove this

File: app/controllers/sessions_controller.rb (hunk 6 above)
Comment (kianaditya) [inline app/controllers/sessions_controller.rb]: Remove comment

File: features/step_definitions/assertion_steps.rb [hunk 11]
```diff
@@ -3,11 +3,15 @@
 end
 
 Then("I should see {string} in {string}") do |session, status|
-	within("##{status}") do
+	within("#{status}") do
```
Comment (GergKllai1) [inline features/step_definitions/assertion_steps.rb]: Why did you change this? Are the trainee can see all sessions tests are passing?

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (kianaditya) [inline app/controllers/sessions_controller.rb]: Why do we need this method?

File: features/step_definitions/assertion_steps.rb [hunk 12]
```diff
@@ -3,11 +3,15 @@
...
-	within("##{status}") do
+	within("#{status}") do 
 		expect(page).to have_content session
 	end
 end
 
 Then("I should not see {string}") do |content|
     expect(page).to have_no_content content
 end
+
```
Comment (GergKllai1) [inline features/step_definitions/assertion_steps.rb]: Remove this we are not using it anymore

File: client/bundles/Main/components/App.jsx [hunk 13]
```diff
@@ -24,6 +25,7 @@ export class App extends React.Component {
 				<Route exact path ='/' render={() => <Main sessions={this.state.sessions} />} />;
 				<Route exact path ='/register' render={() => <Registration />} />;
 				<Route exact path ='/login' render={() => <Login />} />;
+				<Route exact path ='/addsession' render={() => <AddSession />} />;
```
Comment (kianaditya) [inline client/bundles/Main/components/App.jsx]: Just my opinion: instead of url - sessions/addsession, we can use url - sessions/add

**Review by GergKllai1 — CHANGES_REQUESTED** [review]

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: Done

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: We don't, it was connected to the before_action above. Removes this

File: app/controllers/sessions_controller.rb (hunk 6 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: Done

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: done

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: done

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: done

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: done

File: config/routes.rb (hunk 9 above)
Comment (johannaanden) [reply config/routes.rb]: Done

File: config/routes.rb (hunk 10 above)
Comment (johannaanden) [reply config/routes.rb]: Done

File: features/step_definitions/assertion_steps.rb (hunk 11 above)
Comment (johannaanden) [reply features/step_definitions/assertion_steps.rb]: We have changed back to ##{status} to make the test passes

File: features/step_definitions/assertion_steps.rb (hunk 12 above)
Comment (johannaanden) [reply features/step_definitions/assertion_steps.rb]: We are using "I should not see" in the login features

File: client/bundles/Main/components/sessions/AllSessions.jsx (hunk 8 above)
Comment (johannaanden) [reply client/bundles/Main/components/sessions/AllSessions.jsx]: if we don't have the ke we get the folowing error "Each child in an array or iterator should have a unique "key" prop."

File: client/bundles/Main/components/App.jsx (hunk 13 above)
Comment (johannaanden) [reply client/bundles/Main/components/App.jsx]: We will keep addasession to stick to the convention we all ready have on the naming.

File: app/controllers/sessions_controller.rb [hunk 14]
```diff
@@ -1,11 +1,28 @@
...
 			user = current_user
 		end
 		session = Session.find(params[:id])
 		@session_props = {session: session, user: user}
-	end
+    end
+    
+    def create
+        pt = PriceTable.find_by(price_point: session_params[:price_point])
+        @session = Session.new(session_params.except(:price_point).merge(price_table_id: pt.id))
```
Comment (GergKllai1) [inline app/controllers/sessions_controller.rb]: Why are we doing this?

File: client/bundles/Main/components/sessions/AddSession.jsx [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+const axios = require("axios");
+
+export class AddSession extends Component {
+	
+	constructor(props) {
+		super(props);
+		this.state = {
+			title: "",
+			start_date: "",
+			price_point: 1,
```
Comment (kianaditya) [inline client/bundles/Main/components/sessions/AddSession.jsx]: why hard code '1'?

File: db/schema.rb [hunk 16]
```diff
@@ -13,6 +13,7 @@
 ActiveRecord::Schema.define(version: 2019_02_02_172750) do
 
   # These are extensions that must be enabled in order to support this database
+  enable_extension "hstore"
```
Comment (GergKllai1) [inline db/schema.rb]: We are not using this.
Please do rails db:drop create migrate

File: features/support/env.rb [hunk 17]
```diff
@@ -25,6 +25,18 @@
   Capybara::Selenium::Driver.new(app, :browser => :chrome)
 end
 
+# Capybara.register_driver :selenium do |app|
```
Comment (kianaditya) [inline features/support/env.rb]: Do we need this commented code ?

File: features/step_definitions/assertion_steps.rb (hunk 12 above)
Comment (GergKllai1) [reply features/step_definitions/assertion_steps.rb]: I meant the code from line 15:
```
Then("I should be on sessions page") do
	expect(current_path).to eq sessions_path
end
```

File: features/support/env.rb (hunk 17 above)
Comment (GergKllai1) [inline features/support/env.rb]: Do we need this commented out code?

**Review by GergKllai1 — CHANGES_REQUESTED** [review]

File: features/coach_can_create_session.feature [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        And the following user exists
+            | email           | password | password_confirmation | first_name | last_name |
+            | coach@email.com | password | password              | Coach-Jon  | Doe       |
+        And I visit the site
+        And I am logged in as 'coach@email.com'
+        And I click 'Add Session'
+
+    Scenario: Coach can create session
+        Given I fill in 'session[title]' field with 'Spinning'
+        And I fill in 'session[price_point]' field with '1'
```
Comment (kianaditya) [inline features/coach_can_create_session.feature]: Can we use 'Low,medium and high' for price points instead of numbers?

File: features/coach_can_create_session.feature [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+        And I set the date and time
+        And I click 'Add Session'
+        Then I should see 'SPINNING'
+
+    Scenario: Coach can not create session
```
Comment (kianaditya) [inline features/coach_can_create_session.feature:32]: Should we check for trainee cannot create session?

**Review by kianaditya — CHANGES_REQUESTED** [review]

File: app/javascript/css/tailwind.scss (hunk 7 above)
Comment (johannaanden) [reply app/javascript/css/tailwind.scss]: We are not using this class, removed

File: features/step_definitions/assertion_steps.rb (hunk 12 above)
Comment (johannaanden) [reply features/step_definitions/assertion_steps.rb]: done

File: app/controllers/sessions_controller.rb (hunk 14 above)
Comment (johannaanden) [reply app/controllers/sessions_controller.rb]: Because we need to have the price table when we are creating a new session

File: db/schema.rb (hunk 16 above)
Comment (johannaanden) [reply db/schema.rb]: Done

File: features/support/env.rb (hunk 17 above)
Comment (johannaanden) [reply features/support/env.rb]: No, removed

Comment (johannaanden) [reply features/support/env.rb]: No, removed

File: client/bundles/Main/components/sessions/AddSession.jsx (hunk 15 above)
Comment (johannaanden) [reply client/bundles/Main/components/sessions/AddSession.jsx]: We need to refactor our price table to change this, we have created a chore for this in our PT.

File: features/coach_can_create_session.feature (hunk 18 above)
Comment (johannaanden) [reply features/coach_can_create_session.feature]: We need to refactor our price table to change this, we have created a chore for this in our PT.

File: features/coach_can_create_session.feature (hunk 19 above)
Comment (johannaanden) [reply features/coach_can_create_session.feature:32]: Added test for this

**Review by kianaditya — CHANGES_REQUESTED** [review]
Please also resolve Codacy issues

**Review by HannaSwDn — APPROVED** [review]

**Review by kianaditya — APPROVED** [review]

**Review by GergKllai1 — APPROVED** [review]
